tests/frontend/io/test_io_xir.py

```diff
@@ -486,6 +482,36 @@ def test_tdm_program_script(self):
         assert all(sf_prog.tdm_params[1] == np.array([1, 0.5, np.pi]))
         assert all(sf_prog.tdm_params[2] == np.array([0, 0, 0]))
 
+    def test_script_with_gate_definition(self):
+        """Test converting a XIR script with gate definitions to a Program"""
+        xir_script = inspect.cleandoc("""
+            gate Banana(a, b, c, d):
+                Rgate(a) | [0];
+                BSgate(b, c) | [0, 1];
+                Rgate(d) | [1];
+            end;
+
+            Vacuum | [1];
+            Banana(0.5, 0.4, 0.0, 0.5) | [3, 0];
+            Squeezed(0.12, 0.0) | [2];
+            """)
```

Comment:
Nice!

Relatedly, should we raise an exception when an XIR program has an observable declaration or definition?

Reply:
Maybe?! It doesn't really matter much if a user has observables in the script, unless they're used, and if they attempt to use it in any statements the conversion should fail anyway (I think/hope).